validate/validate.go

```diff
@@ -356,9 +356,29 @@ func (v *Validator) CheckLinux() (msgs []string) {
 		msgs = append(msgs, fmt.Sprintf("On Linux, hostname requires a new UTS namespace to be specified as well"))
 	}
 
+	// Linux devices validation
+	devList := make(map[string]bool)
 	for index := 0; index < len(v.spec.Linux.Devices); index++ {
-		if !deviceValid(v.spec.Linux.Devices[index]) {
-			msgs = append(msgs, fmt.Sprintf("device %v is invalid.", v.spec.Linux.Devices[index]))
+		device := v.spec.Linux.Devices[index]
+		if !deviceValid(device) {
+			msgs = append(msgs, fmt.Sprintf("device %v is invalid.", device))
+		}
+		if _, exists := devList[device.Path]; exists {
+			msgs = append(msgs, fmt.Sprintf("device %s is duplicated", device.Path))
+		} else {
+			devList[device.Path] = true
+		}
+		// unify u->c when comparing, they are synonyms
+		var devId string
+		if device.Type == "u" {
+			devId = fmt.Sprintf("%s:%d:%d", "c", device.Major, device.Minor)
+		} else {
+			devId = fmt.Sprintf("%s:%d:%d", device.Type, device.Major, device.Minor)
+		}
+		if _, exists := devList[devId]; exists {
```

Comment:
Now you're using `devList` with two types of keys: `device.Path` and `devId`.  While the risk of collision is low, I'd rather have two separate maps (`devPaths` and `devIds`?) just to be safe.